sync-android/src/main/java/com/cloudant/sync/datastore/encryption/KeyManager.java

```diff
@@ -0,0 +1,217 @@
+
+                byte[] encryptedDpkBytes = DKPEncryptionUtil.encryptAES(aesKey, iv, dpkBytes);
+
+                KeyData keyData = new KeyData(encryptedDpkBytes, salt, iv,
+                        CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS, CDTENCRYPTION_KEYCHAIN_VERSION);
+
+                if (this.storage.saveEncryptionKeyData(keyData)) {
+                    dpk = new EncryptionKey(dpkBytes);
+                }
+            }
+        } catch (Exception e) {
+            throw new DPKException("Failed to encrypt DPK", e);
+        }
+        return dpk;
+    }
+
+    /**
+     * Checks if the encrypted Data Protection Key (DPK) is inside the {@link SharedPreferences}.
+     *
+     * @return true if the encrypted DPK is inside the {@link SharedPreferences}, false otherwise
+     */
+    public boolean keyExists() {
+        return this.storage.encryptionKeyDataExists();
+    }
+
+    /**
+     * Clears security metadata from the {@link SharedPreferences}.
+     *
+     * @return Success (true) or failure (false)
+     */
+    public boolean clearKey() {
+        return this.storage.clearEncryptionKeyData();
+    }
+
+    // PRIVATE HELPER METHODS
+    private boolean validateEncryptionKeyData(KeyData data) {
+        if (data.getIv().length != CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE) {
+            LOGGER.warning("IV does not have the expected size: " +
+                    CDTENCRYPTIONKEYCHAINMANAGER_AES_IV_SIZE + " bytes");
+            return false;
+        }
+        return true;
+    }
+
+    private SecretKey pbkdf2DerivedKeyForPassword(String password, byte[] salt, int
+            iterations, int length) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        if (length < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (password == null || password.length() < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (salt == null || salt.length < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        if (iterations < 1) {
+            throw new IllegalArgumentException("Length must greater than 0");
+        }
+
+        SecretKeyFactory pbkdf2Factory;
+        if (Build.VERSION.SDK_INT >= 19) {
+            // Use compatibility key factory required for backwards compatibility in API 19 and up.
+            pbkdf2Factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1And8bit"); //$NON-NLS-1$
+        } else {
+            // Traditional key factory.
+            pbkdf2Factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1"); //$NON-NLS-1$
+        }
+
+        PBEKeySpec keySpec = new PBEKeySpec(password.toCharArray(), salt, iterations, length *
+                BYTES_TO_BITS); //$NON-NLS-1$
+        return pbkdf2Factory.generateSecret(keySpec);
+    }
+
+    private byte[] generateSecureRandomBytesWithLength(int length) {
```

Comment:
I just moved them internal because they were not actually used by other code. I felt like inclusion in a Util would be justified by reuse. I am open to moving them out if there is a strong preference.